iceoryx_utils/platform/mac/source/semaphore.cpp

```diff
@@ -0,0 +1,253 @@
+        }
+    }
+    else
+    {
+        // dispatch semaphore always succeed
+        dispatch_semaphore_signal(sem->m_handle.dispatch);
+        (*sem->m_value)++;
+    }
+    return retVal;
+}
+
+int iox_sem_wait(iox_sem_t* sem)
+{
+    int retVal{0};
+    if (sem->m_hasPosixHandle)
+    {
+        retVal = sem_wait(sem->m_handle.posix);
+    }
+    else
+    {
+        // dispatch semaphore always succeed
+        dispatch_semaphore_wait(sem->m_handle.dispatch, DISPATCH_TIME_FOREVER);
+    }
+
+    sem->m_value->fetch_sub((retVal == 0) ? 1 : 0);
+    return retVal;
+}
+
+int iox_sem_trywait(iox_sem_t* sem)
+{
+    int retVal{0};
+    if (sem->m_hasPosixHandle)
+    {
+        retVal = sem_trywait(sem->m_handle.posix);
+    }
+    else
+    {
+        if (dispatch_semaphore_wait(sem->m_handle.dispatch, 0) != 0)
+        {
+            errno = EAGAIN;
+            retVal = -1;
+        }
+    }
+
+    sem->m_value->fetch_sub((retVal == 0) ? 1 : 0);
+    return retVal;
+}
+
+int iox_sem_timedwait(iox_sem_t* sem, const struct timespec* abs_timeout)
+{
+    struct timeval tv;
+    gettimeofday(&tv, nullptr);
+
+    static constexpr int64_t NANOSECONDS_PER_SECOND = 1000000000;
+    static constexpr int64_t NANOSECONDS_PER_MICROSECOND = 1000;
+
+    int64_t timeoutInNanoSeconds = std::max(0ll,
+                                            (abs_timeout->tv_sec - tv.tv_sec) * NANOSECONDS_PER_SECOND
+                                                + abs_timeout->tv_nsec - tv.tv_usec * NANOSECONDS_PER_MICROSECOND);
+
+    if (sem->m_hasPosixHandle)
+    {
+        int tryWaitCall = sem_trywait(sem->m_handle.posix);
+        if (tryWaitCall == -1 && errno != EAGAIN)
+        {
+            return -1;
+        }
+        else if (tryWaitCall == -1 && errno == EAGAIN && timeoutInNanoSeconds == 0)
+        {
+            errno = ETIMEDOUT;
+            return -1;
+        }
+        else if (tryWaitCall == 0)
+        {
+            (*sem->m_value)--;
+            return 0;
+        }
+
+        std::this_thread::sleep_for(std::chrono::nanoseconds(timeoutInNanoSeconds));
+
+        tryWaitCall = sem_trywait(sem->m_handle.posix);
+        if (tryWaitCall == -1 && errno == EAGAIN)
+        {
+            errno = ETIMEDOUT;
+            return -1;
+        }
+        else if (tryWaitCall == -1 && errno != EAGAIN)
+        {
+            return -1;
+        }
+        else if (tryWaitCall == 0)
+        {
+            (*sem->m_value)--;
+            return 0;
+        }
+    }
+    else
+    {
+        dispatch_time_t timeout = dispatch_time(DISPATCH_TIME_NOW, timeoutInNanoSeconds);
+        if (dispatch_semaphore_wait(sem->m_handle.dispatch, timeout) != 0)
+        {
+            errno = ETIMEDOUT;
+            return -1;
+        }
+        else
+        {
+            (*sem->m_value)--;
+            return 0;
+        }
+    }
+
+    return -1;
+}
+
+int iox_sem_close(iox_sem_t* sem)
+{
+    // will only be called by named semaphores which are in our case
```

Comment:
/// @note here

Reply:
Developer comment. Not intended for the documentation.